src/renderers/shared/fiber/__tests__/ReactIncrementalScheduling-test.js

```diff
@@ -341,13 +341,60 @@ describe('ReactIncrementalScheduling', () => {
     expect(ReactNoop.getChildren()).toEqual([span(1)]);
   });
 
-  it('can force synchronous updates with syncUpdates, even inside batchedUpdates', done => {
-    ReactNoop.batchedUpdates(() => {
-      ReactNoop.syncUpdates(() => {
-        ReactNoop.render(<span />);
-        expect(ReactNoop.getChildren()).toEqual([span()]);
-        done();
+  it('can opt-out of batching using unbatchedUpdates', () => {
+    // syncUpdates gives synchronous priority to updates
+    ReactNoop.syncUpdates(() => {
+      // batchedUpdates downgrades sync updates to task priority
+      ReactNoop.batchedUpdates(() => {
+        ReactNoop.render(<span prop={0} />);
+        expect(ReactNoop.getChildren()).toEqual([]);
+        // Should not have flushed yet because we're still batching
+
+        // unbatchedUpdates reverses the effect of batchedUpdates, so sync
+        // updates are not batched
+        ReactNoop.unbatchedUpdates(() => {
+          ReactNoop.render(<span prop={1} />);
+          expect(ReactNoop.getChildren()).toEqual([span(1)]);
+          ReactNoop.render(<span prop={2} />);
+          expect(ReactNoop.getChildren()).toEqual([span(2)]);
+        });
+
+        ReactNoop.render(<span prop={3} />);
+        expect(ReactNoop.getChildren()).toEqual([span(2)]);
       });
+      // Remaining update is now flushed
+      expect(ReactNoop.getChildren()).toEqual([span(3)]);
+    });
+  });
+
+  it('nested updates are always deferred, even inside unbatchedUpdates', done => {
+    let instance;
+    class Foo extends React.Component {
+      state = { step: 0 };
+      componentDidUpdate() {
+        if (this.state.step === 1) {
+          ReactNoop.unbatchedUpdates(() => {
+            // This is a nested state update, so it should not be
+            // flushed synchronously, even though we wrapped it
+            // in unbatchedUpdates.
+            this.setState({ step: 2 });
+          });
+          expect(ReactNoop.getChildren()).toEqual([span(1)]);
+          done();
+        }
+      }
+      render() {
+        instance = this;
+        return <span prop={this.state.step} />;
+      }
+    }
+    ReactNoop.render(<Foo />);
+    ReactNoop.flush();
+    expect(ReactNoop.getChildren()).toEqual([span(0)]);
+
+    ReactNoop.syncUpdates(() => {
+      instance.setState({ step: 1 });
+      expect(ReactNoop.getChildren()).toEqual([span(2)]);
```

Comment:
This expectation won't be evaluated. It could fail and Jest won't catch it because you've already called `done` in `componentDidUpdate` above. To resolve this you could set a bool (eg `componentDidUpdateCalled = true`) in `componentDidUpdate` above, verify it here, and then call `done()` at the end of the sync update.

```jsx
it('nested updates are always deferred, even inside unbatchedUpdates', done => {
  let instance;
  let componentDidUpdateCalled = false;
  class Foo extends React.Component {
    state = { step: 0 };
    componentDidUpdate() {
      if (this.state.step === 1) {
        ReactNoop.unbatchedUpdates(() => {
          // This is a nested state update, so it should not be
          // flushed synchronously, even though we wrapped it
          // in unbatchedUpdates.
          this.setState({ step: 2 });
        });
        expect(ReactNoop.getChildren()).toEqual([span(1)]);
        componentDidUpdateCalled = true;
      }
    }
    render() {
      instance = this;
      return <span prop={this.state.step} />;
    }
  }
  ReactNoop.render(<Foo />);
  ReactNoop.flush();
  expect(ReactNoop.getChildren()).toEqual([span(0)]);

  ReactNoop.syncUpdates(() => {
    instance.setState({ step: 1 });
    expect(componentDidUpdateCalled).toBe(true);
    expect(ReactNoop.getChildren()).toEqual([span(2)]);
    done();
  });
});
```